docs/design.md

```diff
@@ -0,0 +1,299 @@
+
+> The indexer reads the advertisement chain starting from the head, reading
+> previous advertisements until a previously seen advertisement, or the end of
+> the chain, is reached. The advertisements and their entries are then processed
+> in order from earliest to head.
+
+An Advertisement has several properties (see
+[the full spec](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#advertisements)),
+we need the following ones:
+
+- **`PreviousID`** is the CID of the previous advertisement, and is empty for
+  the 'genesis'.
+
+- **`Metadata`** represents additional opaque data. The metadata for Graphsync
+  retrievals includes the PieceCID that we are looking for.
+
+- **`Entries`** provide a paginated list of multihashes. For our purposes, it's
+  enough to take the first entry and ignore the rest.
+
+Advertisements are made available for consumption by indexer nodes as a set of
+files fetched via HTTP.
+
+Quoting from
+[Advertisement Transfer](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#advertisement-transfer):
+
+> All IPNI HTTP requests use the IPNI URL path prefix, `/ipni/v1/ad/`. Indexers
+> and advertisement publishers implicitly use and expect this prefix to precede
+> the requested resource.
+>
+> The IPLD objects of advertisements and entries are represented as files named
+> by their CIDs in an HTTP directory. These files are immutable, so can be
+> safely cached or stored on CDNs. To fetch an advertisement or entries file by
+> CID, the request made by the indexer to the publisher is
+> `GET /ipni/v1/ad/{CID}`.
+
+The IPNI instance running at https://cid.contact provides an API returning the
+list of all index providers from which cid.contact have received announcements:
+
+https://cid.contact/providers
+
+The response provides all the metadata we need to download the advertisements:
+
+- `Publisher.Addrs` describes where we can contact SP's index provider
+- `LastAdvertisement` contains the CID of the head advertisement
+
+## Proposed Design
+
+### Ingestion
+
+Ingesting announcements from all Storage Providers is the most complex
+component. For each storage provider, we need to periodically check for the
+latest advertisement head and process the chain from head until we find an
+advertisement we have already processed before. The chain can be very long,
+therefore we need to account for the cases when the service restarts or a new
+head is published before we finish processing the chain.
+
+#### Proposed algorithm
+
+Use the following per-provider state persisted in the database:
+
+- `provider_id` - Primary key.
+
+- `provider_address` - Provider's address where we can fetch advertisements
+  from.
+
+- `last_head` - The CID of the head where we started the previous walk. All
+  advertisements from `last_head` to the end of the chain were already
+  processed.
+
+- `next_head` - The CID of the most recent head seen by cid.contact. This is
```

Comment:
how does `next_head` differ from `head`? Do we not start each walk from the current head?

Reply:
The initial walk will take a long time to complete. While we are walking the "old" chain, new advertisements (new heads) will be announced to IPNI.

- `next_head` is the latest head announced to IPNI
- `head` is the advertisement where the current walk-in-progress started


I suppose we don't need to keep track of `next_head`. When the current walk finishes, we will wait up to one minute until we make another request to cid.contact to find what are the latest heads for each SPs.

In my current proposal, when the current walk finishes, we can immediately continue with walking from the `next_head`.

Reply:
See also the diagram in https://github.com/filecoin-station/piece-indexer/pull/2/files#r1689884939

Reply:
I captured my explanation in the design doc.